### PR TITLE
apply ownerReference to ServingRuntime related resources

### PR DESCRIPTION
**Pull request by christianvogt**
Closes: https://github.com/opendatahub-io/odh-dashboard/issues/1998

## Description

When creating a `ServingRuntime`, adds owner references to related resources:
- `RoleBinding`
- `ServiceAccount`
- ~`Secrets`~ _secrets are already associated with the `ServiceAccount` and get removed automatically_

When deleting a `ServingRuntime`, for compatibility in existing areas, we will continue to attempt to delete the service account and role binding.

In support of kserve, when an `InferenceService` is deleted from the global model serving page, the referenced `ServingRuntime` is also deleted (manually, not by use of owner references).

## How Has This Been Tested?

- Create a model server with all form options.
- Deploy a model to the new server.
- Inspect newly added `RoleBinding`, `ServiceAccount`, `Secrets` for `ownerReferences` pointing back to the previously created `ServingRuntime`.
- Delete the `ServingRuntime` using the UI and ensure all the related resources are removed.
- Now perform the exact same steps again using all the same name values previously entered.
- Delete the `ServingRuntime` from the CLI and ensure all related resources are removed.

## Test Impact

Added tests to utilities.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

**Comment by strangiato** [comment]
I think there may be a fundamental misunderstanding of what Owner references are for here.

An owner reference is a way that an operator is able to indicate that specific children objects that are created by an operand are owned and managed by that specific operand.  

For example, the DSP operator manages the DSPA objects.  When creating a DSPA, the DSP operator will create several deployments.  The DSP operator should place an owner reference on each of those deployments indicating that they are owned by that specific DSPA object.

In the case of the ServingRuntime, you want to add an Owner Reference to the RoleBinding, Service Account, and Secret.  If the Model Mesh controller is creating these objects, the Model Mesh Controller is responsible for setting the owner reference on these objects, NOT the dashboard.  If it is not doing that today and they need to be set, this should be an issue that is opened on the model mesh controller.

If the dashboard is creating these objects and you think they should be owned by the ServingRuntime, that would indicate that the dashboard probably shouldn't be creating these objects in the first place.  The model mesh controller should be enhanced to be able to create and manage these objects with the creation of  a ServingRuntime in those cases.

If the model mesh controller should not be creating these resources and the dashboard is truly the correct tool to be creating them, they are not owned by the ServingRuntime and the dashboard should not be setting an OwnerReference on these objects.

No matter what, the dashboard should not ever be setting an OwnerReference on an object.  It is not an operator that owns and manages objects.  Setting the owner reference manually can and will cause problems and confusion.

If your goal is simply to associate objects together to use allow for easier cleanup, you should be setting labels on those objects.

**Comment by lucferbux** [comment]
Hi @strangiato thanks for the feedback here. I thought any object can be Owner/Dependant regardless of the Operator pattern.
[Kubernetes docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/) point at any object, so I think it's fine to use it in the way we intend to use them. 
There's even a section about [garbace collection](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#owners-dependents) in which they tell you that you can use owners and dependants for deletion.

Let me know if that's not ok so we can take a deeper look. Thanks again for the heads up.

File: frontend/src/api/k8s/inferenceServices.ts [hunk 1]
```diff
@@ -23,37 +24,40 @@ const assembleInferenceService = (
...
   const dataConnectionKey = secretKey || dataConnection;
 
-  return {
-    apiVersion: 'serving.kserve.io/v1beta1',
-    kind: 'InferenceService',
-    metadata: {
-      name,
-      namespace: project,
-      labels: {
+  return addOwnerReference(
```
Comment (lucferbux) [inline frontend/src/api/k8s/inferenceServices.ts:27]: Just to sync this with you, I assume you are expecting that we add the same in `assembleServingRuntime` for kserve right? So we have a circular reference just for that platform right?

**Review by lucferbux — COMMENTED** [review]
/lgtm

Not approving since we have an open discussion, but tested it and so far looks great, I assume we'll need a follow up for kserve, I'll adapt this and get it into my pr!

**Comment by codekow** [comment]
@strangiato Did a great job at explaining the purpose of `ownerReference`. In other mature operators the pattern he describes is very established.

In short... The dashboard should NEVER be the operator or owner of objects.

TL;DR
- The dashboard should NOT contain logic to deploy resources 
  - The dashboard can aid the user in creating CRs but the Operator should render / deploy resources from a CR.
- Maybe consider the dashboard as a client / front end.

File: frontend/src/api/k8s/inferenceServices.ts (hunk 1 above)
Comment (christianvogt) [reply frontend/src/api/k8s/inferenceServices.ts]: I've removed the owner reference on inference service for now. So no we will not need a change to `assumbleServingRuntime` right now.

**Comment by christianvogt** [comment]
@strangiato Thanks for your input. I've done some further research and discussed your concerns with architecture council. There doesn't seem to be a consensus that owner references are strictly a mechanism for use by operator / controllers. They are in support of garbage collection. In this particular case the dashboard is creating resources in response to a user initiated action to create a `ServingRuntime` with auth enabled. The lifespan of these resources are tied to that of the `ServingRuntime` and therefore we expect these resources to be deleted when the `ServingRuntime` is deleted. The use of owner references here suits the use case. That being said, the change now limits the use of owner references to the service account and role binding used for auth.

**Review by lucferbux — APPROVED** [review]
/lgtm

Tested and it's working, I'm gonna adapt the code in my PR

**Comment by lucferbux** [comment]
@christianvogt Agh, I dint' realize it didn't have the latest changes, let's wait then

**Review by lucferbux — APPROVED** [review]
/lgtm

Tested, all the files are bein deleted as intended
